RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitProducerGPU.cc

```diff
@@ -0,0 +1,496 @@
+    setup.get<EcalTimeCalibConstantsRcd>().get(timeCalibConstantsHandle_);
+    setup.get<EcalSampleMaskRcd>().get(sampleMaskHandle_);
+    setup.get<EcalTimeOffsetConstantRcd>().get(timeOffsetConstantHandle_);
+
+    auto const& pedProduct = pedestalsHandle_->getProduct(ctx.stream());
+    auto const& gainsProduct = gainRatiosHandle_->getProduct(ctx.stream());
+    auto const& pulseShapesProduct = pulseShapesHandle_->getProduct(ctx.stream());
+    auto const& pulseCovariancesProduct = pulseCovariancesHandle_->getProduct(ctx.stream());
+    auto const& samplesCorrelationProduct = samplesCorrelationHandle_->getProduct(ctx.stream());
+    auto const& timeBiasCorrectionsProduct = timeBiasCorrectionsHandle_->getProduct(ctx.stream());
+    auto const& timeCalibConstantsProduct = timeCalibConstantsHandle_->getProduct(ctx.stream());
+
+    // bundle up conditions
+    ecal::multifit::ConditionsProducts conditions {
+        pedProduct, gainsProduct, pulseShapesProduct,
+        pulseCovariancesProduct, 
+        samplesCorrelationProduct,
+        timeBiasCorrectionsProduct,
+        timeCalibConstantsProduct,
+        *sampleMaskHandle_,
+        *timeOffsetConstantHandle_,
+        timeCalibConstantsHandle_->getOffset()
+    };
+    
+    //
+    // retrieve collections
+    //
+    edm::Handle<EBDigiCollection> ebDigis;
+    edm::Handle<EEDigiCollection> eeDigis;
+    event.getByToken(digisTokenEB_, ebDigis);
+    event.getByToken(digisTokenEE_, eeDigis);
+
+    ecal::multifit::EventInputDataCPU eventInputDataCPU{*ebDigis, *eeDigis};
+
+    //
+    // schedule algorithms
+    //
+    ecal::multifit::entryPoint(
+        eventInputDataCPU,
+        eventInputDataGPU_,
+        eventOutputDataGPU_,
+        eventDataForScratchGPU_,
+        conditions,
+        configParameters_,
+        ctx.stream()
+    );
+        
+    ebRecHits_ = std::move(
+        std::make_unique<ecal::UncalibratedRecHit<ecal::Tag::soa>>());
+    eeRecHits_ = std::move(
+        std::make_unique<ecal::UncalibratedRecHit<ecal::Tag::soa>>());
+
+    if (shouldTransferToHost_) {
+        // allocate for the result while kernels are running
+        ebRecHits_->resize(ebDigis->size());
+        eeRecHits_->resize(eeDigis->size());
+
+        // det ids are host copy only - no need to run device -> host
+        std::memcpy(ebRecHits_->did.data(), ebDigis->ids().data(), 
+            ebDigis->ids().size() * sizeof(uint32_t));
+        std::memcpy(eeRecHits_->did.data(), eeDigis->ids().data(),
+            eeDigis->ids().size() * sizeof(uint32_t));
+    }
+}
+
+void EcalUncalibRecHitProducerGPU::produce(
+        edm::Event& event, 
+        edm::EventSetup const& setup) 
+{
+    DurationMeasurer<std::chrono::milliseconds> timer{std::string{"produce duration"}};
+    CUDAScopedContextProduce ctx{cudaState_};
+
+    if (shouldTransferToHost_) {
+        // rec hits objects were not originally member variables
+        transferToHost(*ebRecHits_, *eeRecHits_, ctx.stream());
```

Comment:
These transfers could be queued already in `acquire()`, then the `cudaStreamSynchronize()` below could be removed.

Reply:
On the other hand, the target memory is non-pinned (unless I missed something), so the transfer is synchronous anyway.